lib/resumption/resumable.go

```diff
@@ -0,0 +1,339 @@
+			}
+
+			if maxAck := r.sendBuffer.len(); ack > maxAck {
+				r.mu.Unlock()
+				return trace.BadParameter("got ack bigger than current send buffer (%v, expected up to %v)", ack, maxAck)
+			}
+
+			r.sendBuffer.advance(ack)
+			r.cond.Broadcast()
+			r.mu.Unlock()
+		}
+
+		size, err := binary.ReadUvarint(nc)
+		if err != nil {
+			return trace.Wrap(err, "reading data size")
+		}
+
+		if size > maxFrameSize {
+			return trace.BadParameter("got data size bigger than limit (%v, expected up to %v)", size, maxFrameSize)
+		}
+
+		r.mu.Lock()
+
+		for size > 0 {
+			if r.localClosed {
+				r.receiveBuffer.advance(r.receiveBuffer.len())
+				r.cond.Broadcast()
+				r.mu.Unlock()
+
+				n, err := io.Copy(io.Discard, io.LimitReader(nc, int64(size)))
+
+				r.mu.Lock()
+				r.receiveBuffer.start += uint64(n)
+				r.receiveBuffer.end = r.receiveBuffer.start
+				r.cond.Broadcast()
+				if err != nil {
+					r.mu.Unlock()
+					return trace.Wrap(err, "reading data to discard")
+				}
+				break
+			}
+
+			for r.receiveBuffer.len() >= receiveBufferSize {
+				r.cond.Wait()
+				if stopRequested.Load() || r.remoteClosed {
+					r.mu.Unlock()
+					if stopRequested.Load() {
+						return trace.ConnectionProblem(net.ErrClosed, "disconnection requested")
+					}
+					return trace.ConnectionProblem(net.ErrClosed, "connection closed by peer")
+				}
+			}
+
+			next := min(receiveBufferSize-r.receiveBuffer.len(), size)
+			r.receiveBuffer.reserve(next)
+			tail, _ := r.receiveBuffer.free()
+			if len64(tail) > size {
+				tail = tail[:size]
+			}
+			r.mu.Unlock()
+
+			n, err := io.ReadFull(nc, tail)
+
+			r.mu.Lock()
+			if n > 0 {
+				r.receiveBuffer.append(tail[:n])
+				size -= uint64(n)
+				r.cond.Broadcast()
+			}
+
+			if err != nil {
+				r.mu.Unlock()
+				return trace.Wrap(err, "reading data")
+			}
+		}
+		r.mu.Unlock()
+	}
+}
+
+func runResumeV1Write(r *Conn, nc io.Writer, stopRequested *atomic.Bool, sentPosition, peerPosition uint64) error {
+	var headerBuf [2 * binary.MaxVarintLen64]byte
+	var dataBuf [maxFrameSize]byte
+
+	for {
+		var frameAck uint64
+		var frameData []byte
+
+		r.mu.Lock()
+		for {
+			frameAck = r.receiveBuffer.end - sentPosition
+
+			frameData = nil
+			if r.sendBuffer.end > peerPosition {
+				skip := peerPosition - r.sendBuffer.start
+				d1, d2 := r.sendBuffer.buffered()
+				if len64(d1) <= skip {
+					frameData = d2[skip-len64(d1):]
+				} else {
+					frameData = d1[skip:]
+				}
+			}
+			if frameAck > 0 || len(frameData) > 0 {
+				break
+			}
+
+			if stopRequested.Load() {
+				r.mu.Unlock()
+				return trace.ConnectionProblem(nil, "disconnection requested")
+
+			}
+
+			if r.remoteClosed {
+				r.mu.Unlock()
+				return trace.ConnectionProblem(errBrokenPipe, "connection closed by peer")
+			}
+
+			if r.localClosed {
+				r.mu.Unlock()
+
+				_, _ = nc.Write([]byte(errorTagUvarint))
+				return trace.ConnectionProblem(net.ErrClosed, "connection closed")
```

Comment:
Won't `errors.Is` also unwrap the error if necessary? I don't think anything expects either, so we can wrap a `nil` error with `ConnectionProblem` or just `trace.Wrap` the semantic errors with additional messages.

Reply:
Replaced every synthesized error with a `trace.Wrap(net.ErrClosed, ...)`, since that will get ignored when logging (by going through `utils.IsOKNetworkError`) and the only important errors that we might care about (as they'd reveal logical problems) are still `trace.BadParameter` anyway.